### PR TITLE
TimedBuffer: Buffer for key value pairs with an expiration time.

### PR DESCRIPTION
**Pull request by brian-js**
Adds a timed buffer struct for use with requests that require waiting for a response.

+@dirvine +@qima +@Fraser999

---
This change is [Reviewable](https://reviewable.io/reviews/maidsafe/safe_vault/385)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @Fraser999

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/timed_buffer.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /// Removes a value from the buffer.
+    pub fn remove(&mut self, key: &Key) -> Option<Value> {
+        self.map.remove(key).map(|(value, _)| value)
+    }
+
+    /// Retrieves a reference to the value stored under `key`, or `None` if the key doesn't exist.
+    pub fn get(&mut self, key: &Key) -> Option<Value> {
+        self.map.get(key).map(|&(ref value, _)| value.clone())
```
Comment (afck) [inline src/timed_buffer.rs]: This should return a reference, not clone the value, as the documentation says.

Comment (brian-js) [inline src/timed_buffer.rs]: @afck It's a TimedBuffer and the documentation has nothing to say about how this gets returned. I think it was also agreed in the vault channel to do this.

**Comment by afck** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/timed_buffer.rs, line 51 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDd2VWDD30wvvuH7sdj-r1-51:-KDd8w4K48ccunaWzIOT:657909747) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L51)):*
I mean the method documentation, which says that a reference is returned. That should probably be rephrased then.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by brian-js** [comment]
*[src/timed_buffer.rs, line 51 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDd2VWDD30wvvuH7sdj-r1-51:-KDd9fa0bFv9zJmq8DBf:361520655) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L51)):*
Your right. I'll change it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by maqi** [comment]
*[a discussion](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDdA6HzFJT6a8C20gTR:-KDdA6HzFJT6a8C20gTS:1616557553) (no related file):*
so updating PM to use this time_buffer will be in a separate PR?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by brian-js** [comment]
*[a discussion](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDdA6HzFJT6a8C20gTR:-KDdAGT6lyY2HkwHDHwc:677251276) (no related file):*
Yes.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by maqi** [comment]
*[src/timed_buffer.rs, line 41 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDdBF0ostjgaoUwCtYO:-KDdBF0p2_WbpeD6qBYi:-1820261682) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L41)):*
in case None is returned from  BTreeMap.insert(), will calling .map(|pair| pair.0) raising a panic?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by brian-js** [comment]
*[src/timed_buffer.rs, line 41 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDdBF0ostjgaoUwCtYO:-KDdH9irr5AmuuGnFDhB:1758359711) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L41)):*
Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by maqi** [comment]
Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by Fraser999** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/timed_buffer.rs, line 51 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDd2VWDD30wvvuH7sdj-r1-51:-KDdK3P837dmCfRvS2JK:472071671) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L51)):*
The discussion in the Vault channel mentioned popping the item and returning that, not copying it.  It looks like that discussion was more applicable to `remove`.  I think it would be more normal (and efficient) to return a ref here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by dirvine** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/timed_buffer.rs, line 51 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDd2VWDD30wvvuH7sdj-r1-51:-KDdNX-vovjeNjrWhRqT:1171500925) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L51)):*
Would it be best if the value (if any) was moved out of the vector and returned. If we get it and it's the only one we get then perhaps it does. If we Get multiple times then a reference or copy if we want ownership to stay with vector (which seems to be a btree). So I suppose we need to answer does a Get mean it's done and can only be done once? first. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by Fraser999** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/timed_buffer.rs, line 51 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDd2VWDD30wvvuH7sdj-r1-51:-KDdQdTJ3XQj5_g50qJc:679961661) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L51)):*
I guess if we define `get` that way, we don't want `remove` as well, since they'd be identical?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by brian-js** [comment]
*[src/timed_buffer.rs, line 51 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDd2VWDD30wvvuH7sdj-r1-51:-KDde5oLXd619eeaoExU:413908731) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L51)):*
Removed unrequired get.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by maqi** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by brian-js** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/timed_buffer.rs, line 51 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/385#-KDd2VWDD30wvvuH7sdj-r1-51:-KDdz0iHJK27d33B70TR:-499146009) ([raw file](https://github.com/maidsafe/safe_vault/blob/31450e36a9a388ac11e6ab61dacb10a18cb70f88/src/timed_buffer.rs#L51)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by brian-js** [comment]
Reviewed 1 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Comment by dirvine** [comment]
Reviewed 1 of 2 files at r1, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_vault/385)*

**Pull request merged**